nixos/modules/programs/wayland/niri.nix

```diff
@@ -0,0 +1,90 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.programs.niri;
+in
+{
+  options.programs.niri = {
+    enable = lib.mkEnableOption "Niri, a scrollable-tiling Wayland compositor";
+
+    package = lib.mkPackageOption pkgs "niri" { };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#authentication-agent
+    polkitAgent.enable = lib.mkEnableOption "a PolicyKit agent service" // {
+      default = config.security.polkit.enable;
+      defaultText = "config.security.polkit.enable";
+    };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+    portalPackage = lib.mkPackageOption pkgs "xdg-desktop-portal-gnome" {
+      # https://github.com/NixOS/nixpkgs/pull/348193#discussion_r1798515394
+      extraDescription = "Note: `xdg-desktop-portal-gnome` is required for screencast support";
+    };
+
+    xwayland.enable = lib.mkEnableOption "XWayland support via xwayland-satellite";
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      {
+        environment.systemPackages = [ cfg.package ];
+
+        services = {
+          displayManager.sessionPackages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+          gnome.gnome-keyring.enable = lib.mkDefault true;
+        };
+
+        systemd.packages = [ cfg.package ];
```

Comment:
(see also previous revision discussion)

while i understand now that this option isn't actually a no-op, is it not at best still redundant with `environment.systemPackages`? that option definitely adds the relevant user services, so adding the package here as well doesn't seem to have any net effect in the end. why are we doing it?

Reply:
> It looks more complete like this and it makes sure that any additional logic that gets added to the nixos systemd stuff, will also take niri into account

This is also fairly common with modules from what I've seen

Reply:
> it makes sure that any additional logic that gets added to the nixos systemd stuff, will also take niri into account

yeah ok, fair enough. i agree that this is probably more correct.

> This is also fairly common with modules from what I've seen

checks out. from a glance, the first few results using this are a close split between "isn't actually installed in `environment.systemPackages`, but adds the service using `systemd.packages`" (e.g. [stratis](https://github.com/NixOS/nixpkgs/blob/bd95a9ab27e843d42e6bc6b6bebf5587b460b6e8/nixos/modules/tasks/stratis.nix#L12-L13)) and "present in both" (e.g. [gpaste](https://github.com/NixOS/nixpkgs/blob/bd95a9ab27e843d42e6bc6b6bebf5587b460b6e8/nixos/modules/programs/gpaste.nix#L21-L23))

---

acceptable.